replica/database.cc

```diff
@@ -2250,6 +2251,10 @@ future<> database::shutdown() {
     co_await _stop_barrier.arrive_and_wait();
     b.cancel();
 
+    // quiesce compaction across all shards before closing tables
+    co_await _compaction_manager.drain();
+    co_await _stop_barrier.arrive_and_wait();
```

Comment:
I think that's still not enough. we need to drain compaction on all shards first, and only then call database::stop().

Otherwise it could happen:

1) shard 1 calls compaction_manager::drain() and closes compaction_history table
2) shard 2 completes compaction and tries to append a change to compaction_history in shard 1.

That will fail due to cross-shard communication.

What we need to do is to call _compaction_manager.drain() on all shards, and only then proceed to database::stop()

The change will be similar to what I suggested:

diff --git a/main.cc b/main.cc
index 3687ef0907..e6cd28f401 100644
--- a/main.cc
+++ b/main.cc
@@ -1056,6 +1056,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
                 // #293 - do not stop anything - not even db (for real)
                 //return db.stop();
                 // call stop on each db instance, but leave the shareded<database> pointers alive.
+                db.invoke_on_all(&replica::database::quiesce_compaction).get();
                 db.invoke_on_all(&replica::database::stop).get();
             });

replica::database::quiesce_compaction is simply a call to _compaction_manager.drain()

Reply:
@raphaelsc Okay I'll update the code.

My assumption was `co_await _stop_barrier.arrive_and_wait()` will protect the us from the race condition as it makes all the shards wait until `_compaction_manager.drain()` has succeeded in all the shards. Is that not enough?

Reply:
Ah sorry about that, you used a cross shard barrier I wasn't aware of.

Reply:
Yes, that should work!

Reply:
Okay.

Reply:
@raphaelsc done.